### PR TITLE
Make LocalStore an interface.

### PR DESCRIPTION
**Pull request by wu-hui**
Make LocalStore an interface and have `LocalStoreImpl implements LocalStore`. Public fields in `LocalStoreImpl` then becomes file private, because `LocalStoreImpl` is not exported.

### CODE REVIEW

File: packages/firestore/src/local/local_store.ts [hunk 1]
```diff
@@ -139,7 +139,142 @@ export interface QueryResult {
...
+  executeQuery(query: Query, usePreviousResults: boolean): Promise<QueryResult>;
+
+  collectGarbage(garbageCollector: LruGarbageCollector): Promise<LruResults>;
+}
+
+/**
+ * Implements `LocalStore` interface.
+ *
+ * Note: some field defined in this class might have public access level, but
+ * the class is not exported so they are only accessible from this file. This is
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_store.ts:273]: s/file/module (the JS specific term)

File: packages/firestore/src/local/local_store.ts [hunk 2]
```diff
@@ -139,7 +139,142 @@ export interface QueryResult {
...
+  /**
+   * Remove mutations from the MutationQueue for the specified batch;
+   * LocalDocuments will be recalculated.
+   *
+   * @returns The resulting modified documents.
+   */
+  rejectBatch(batchId: BatchId): Promise<MaybeDocumentMap>;
+
+  /**
+   * Returns the largest (latest) batch id in mutation queue that is pending server response.
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_store.ts:183]: This line wrap looks off.

File: packages/firestore/src/local/local_store.ts [hunk 3]
```diff
@@ -139,7 +139,142 @@ export interface QueryResult {
...
+
+/**
+ * Implements `LocalStore` interface.
+ *
+ * Note: some field defined in this class might have public access level, but
+ * the class is not exported so they are only accessible from this file. This is
+ * useful to implement optional features (like bundles) in free functions, such
+ * that they are tree-shakeable.
+ */
+class LocalStoreImpl implements LocalStore {
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_store.ts:277]: Do you mind removing the JSDocs that you copied over? It is too easy for them to go out of sync now that they are in two places.

File: packages/firestore/src/local/local_store.ts [hunk 4]
```diff
@@ -979,12 +1114,59 @@ export class LocalStore {
...
+   */
+  synchronizeLastDocumentChangeReadTime(): Promise<void>;
+}
+
 /**
  * An implementation of LocalStore that provides additional functionality
  * for MultiTabSyncEngine.
+ *
+ * Note: some field defined in this class might have public access level, but
+ * the class is not exported so they are only accessible from this file. This is
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_store.ts:1163]: s/file/module

Please also remove duplicate comments for `MultiTabLocalStoreImpl`.

File: packages/firestore/src/local/local_store.ts (hunk 2 above)
Comment (wu-hui) [reply packages/firestore/src/local/local_store.ts]: Fixed.

File: packages/firestore/src/local/local_store.ts (hunk 1 above)
Comment (wu-hui) [reply packages/firestore/src/local/local_store.ts]: Done.

File: packages/firestore/src/local/local_store.ts (hunk 3 above)
Comment (wu-hui) [reply packages/firestore/src/local/local_store.ts:280]: Done.

File: packages/firestore/src/local/local_store.ts (hunk 4 above)
Comment (wu-hui) [reply packages/firestore/src/local/local_store.ts]: Done.

**Pull request merged**